"Double Write" Identity Allocation Mode for seamless KVStore -> CRD identity migration (#31920)
cilium-team/cilium: Documentation/operations/upgrade.rst

----------------------------------------------------------------------------

Beginning with Cilium 1.6, Kubernetes CRD-backed security identities can be
used for smaller clusters. Along with other changes in 1.6, this allows
kvstore-free operation if desired. It is possible to migrate identities from an
existing kvstore deployment to CRD-backed identities. This minimizes
disruptions to traffic as the update rolls out through the cluster.

Migration
~~~~~~~~~

When identities change, existing connections can be disrupted while Cilium
initializes and synchronizes with the shared identity store. The disruption
occurs when new numeric identities are used for existing pods on some instances
and others are used on others. When converting to CRD-backed identities, it is
possible to pre-allocate CRD identities so that the numeric identities match
those in the kvstore. This allows new and old Cilium instances in the rollout
to agree.

There are two ways to achieve this: you can either run a one-off ``cilium preflight migrate-identity`` script
which will perform a point-in-time copy of all identities from the kvstore to CRDs (added in Cilium 1.6), or use the "Double Write" identity

> allocation mode which will have Cilium manage identities in both the kvstore and CRD at the same time for a seamless migration (added in Cilium 1.17).

Reviewer: Is `added in Cilium 1.17` accurate? I think that these changes will land in a patch release of 1.16 before they land in 1.17.

Author: I am not sure tbh but Marco suggested 1.17 here: https://github.com/cilium/cilium/pull/31920#discussion_r1660993202

There are still a couple of unresolved issues in this PR so I am not sure how long all of it will take.

Reviewer 2: > I think that these changes will land in a patch release of 1.16 before they land in 1.17.

Hmm, not sure I'm following. This PR will be part of v1.17, as it missed the v1.16 feature freeze some weeks ago, and being a new feature it doesn't qualify for backport to stable branches.

Reviewer: Ah I see, I got confused. Thanks for clarifying!

Migration with the ``cilium preflight migrate-identity`` script
###############################################################

The ``cilium preflight migrate-identity`` script is a one-off tool that can be used to copy identities from the kvstore into CRDs.
It has a couple of limitations:

> * If any identity is created in the kvstore after the one-off migration has been completed, it will not be copied into CRDs.

Reviewer: ```suggestion
* If an identity is created in the kvstore after the one-off migration has been completed, it will not be copied into a CRD.
```

  This means that you need to perform the migration on a cluster with no identity churn.
* There is no easy way to revert back to ``--identity-allocation-mode=kvstore`` if something goes wrong after

>   the Agents have been migrated to ``--identity-allocation-mode=crd``

Reviewer: ```suggestion
  Cilium has been migrated to ``--identity-allocation-mode=crd``.
```

If these limitations are not acceptable, it is recommended to use the ":ref:`Double Write <double_write_migration>`" identity allocation mode instead.

> The steps below show an example of the migration using the ``cilium preflight migrate-identity`` script.

Reviewer: ```suggestion
The following steps show an example of performing the migration using the ``cilium preflight migrate-identity`` script.
```

It is safe to re-run the command if desired. It will identify already allocated identities or ones that
cannot be migrated. Note that identity ``34815`` is migrated, ``17003`` is
already migrated, and ``11730`` has a conflict and a new ID allocated for those
labels.

The steps below assume a stable cluster with no new identities created during

> the rollout. Once a Cilium using CRD-backed identities is running, it may begin

Reviewer: ```suggestion
the rollout. Once Cilium using CRD-backed identities is running, it may begin
```

Clearing CRD identities
~~~~~~~~~~~~~~~~~~~~~~~

If a migration has gone wrong, it possible to start with a clean slate. Ensure that no Cilium instances are running with ``--identity-allocation-mode=crd`` and execute:

.. code-block:: shell-session

      $ kubectl delete ciliumid --all

.. _double_write_migration:

> Migration with the "Double Write" identity allocation mode

Reviewer: The docs here describe a KVStore -> CRD migration, but couldn't this technically also be used for a CRD -> KVStore migration?

Author: Yeah, it's possible to do this as well! I added an explicit note about it in the docs.

##########################################################

.. include:: ../beta.rst

The "Double Write" Identity Allocation Mode allows Cilium to allocate identities as KVStore values *and* as CRDs at the

> same time. This mode also has two versions (``--identity-allocation-mode=doublewrite-readkvstore`` and ``--identity-allocation-mode=doublewrite-readcrd``)
> which allow controlling whether reads happen from KVStore or from CRDs.

Reviewer: ```suggestion
same time. This mode also has two versions: one where the source of truth comes from the kvstore, and one where the source of truth comes from CRDs .
```

.. note::

>     The "Double Write" mode is not compatible with Consul as the KVStore

Reviewer: ```suggestion
    "Double Write" mode is not compatible with Consul as the KVStore
```

The high-level migration plan looks as follows:

> 
> #. Starting state: Cilium is running in KVStore mode
> #. Switch Cilium to “Double Write” mode with all reads happening from the KVStore still. This is almost the same as the
>    pure KVStore mode with the only difference being that all identities are duplicated as CRDs but are not used.
> #. Switch Cilium to “Double Write” mode with all reads happening from CRDs. This is equivalent to Cilium running in
>    pure CRD mode but identities will still be updated in the KVStore for possibility of fast rollback
> #. Switch Cilium to CRD mode. The KVStore will no longer be used and will be ready for decommission

Reviewer: ```suggestion

#. Starting state: Cilium is running in KVStore mode.
#. Switch Cilium to “Double Write” mode with all reads happening from the KVStore. This is almost the same as the
   pure KVStore mode with the only difference being that all identities are duplicated as CRDs but are not used.
#. Switch Cilium to “Double Write” mode with all reads happening from CRDs. This is equivalent to Cilium running in
   pure CRD mode but identities will still be updated in the KVStore to allow for the possibility of a fast rollback.
#. Switch Cilium to CRD mode. The KVStore will no longer be used and will be ready for decommission.
```

> This will allow you to perform a gradual and seamless migration with possibility of fast rollback at steps 2 and 3.

Reviewer: ```suggestion
This will allow you to perform a gradual and seamless migration with the possibility of a fast rollback at steps two or three.
```

> Furthermore, when the "Double Write" mode is enabled, the Operator will emit additional metrics which help with monitoring the
> migration progress and can be used for alerting about identity inconsistencies between KVStore and CRDs.

Reviewer: ```suggestion
Furthermore, when the "Double Write" mode is enabled, the Operator will emit additional metrics to help monitor the
migration progress. These metrics can be used for alerting about identity inconsistencies between the KVStore and CRDs.
```

Rollout Instructions
~~~~~~~~~~~~~~~~~~~~

> #. Re-deploy the Operator and the Agents with ``--identity-allocation-mode=doublewrite-readkvstore``.

Reviewer: Is there a specific order to the rollout between the Operator and the Agents?

Author: There is no strict requirement but it's actually better to switch the Operator to Double Write mode first and then keep it this way until the Agents are fully migrated. This is because the Operator does GC so you don't want to be in a situation where the Agents write identities but they are not being cleaned up.
I updated the docs to reflect that.

#. Monitor the Operator metrics and logs to ensure that all identities have converged between the KVStore and CRDs. The relevant metrics emitted by the Operator are:

>    * ``cilium_operator_identity_crd_total_count`` and ``cilium_operator_identity_kvstore_total_count`` reporting the total number of identities in CRDs and KVStore respectively.
>    * ``cilium_operator_identity_crd_only_count`` and ``cilium_operator_identity_kvstore_only_count`` reporting the number of

Reviewer: ```suggestion
   * ``cilium_operator_identity_crd_total_count`` and ``cilium_operator_identity_kvstore_total_count`` report the total number of identities in CRDs and KVStore respectively.
   * ``cilium_operator_identity_crd_only_count`` and ``cilium_operator_identity_kvstore_only_count`` report the number of
```

>      identities that are only in CRDs or KVStore respectively. These metrics should help with detecting inconsistencies.

Reviewer: ```suggestion
     identities that are only in CRDs or only in the KVStore respectively, to help detect inconsistencies.
```

   In case further investigation is needed, the Operator logs will contain detailed information about the discrepancies between KVStore and CRD identities.

>    Note that Garbage Collection for KVStore identities and CRD identities happens at slightly different times, so it is possible to see some discrepancies in the metrics
>    for some periods of time (depending on ``--identity-gc-interval`` and ``--identity-heartbeat-timeout`` settings)

Reviewer: ```suggestion
   Note that Garbage Collection for KVStore identities and CRD identities happens at slightly different times, so it is possible to see discrepancies in the metrics
   for certain periods of time, depending on ``--identity-gc-interval`` and ``--identity-heartbeat-timeout`` settings.
```

#. Once all identities have converged, re-deploy the Operator and the Agents with ``--identity-allocation-mode=doublewrite-readcrd``.

>    This will make Cilium read identities only from CRDs but still write them to KVStore.

Reviewer: ```suggestion
   This will cause Cilium to read identities only from CRDs, but continue to write them to the KVStore.
```

which will perform a point-in-time copy of all identities from the kvstore to CRDs (added in Cilium 1.6), or use the "Double Write" identity
allocation mode which will have Cilium manage identities in both the kvstore and CRD at the same time for a seamless migration (added in Cilium 1.16).

Migration with the ``cilium preflight migrate-identity`` script
###############################################################

The ``cilium preflight migrate-identity`` script is a one-off tool that can be used to copy identities from the kvstore into CRDs.
It has a couple of limitations:

* If any identity is created in the kvstore after the one-off migration has been completed, it will not be copied into CRDs.
  This means that you need to perform the migration on a cluster with no identity churn.
* There is no easy way to revert back to ``--identity-allocation-mode=kvstore`` if something goes wrong after
  the Agents have been migrated to ``--identity-allocation-mode=crd``

If these limitations are not acceptable, it is recommended to use the ":ref:`Double Write <double_write_migration>`" identity allocation mode instead.

> The steps below show an example of such a migration. It is safe to re-run the

Reviewer 2: I'd suggest to change "such a migration" with something more explicit (maybe the "one-off migration"?), as possibly confusing now with the additions above.

----------------------------------------------------------------------------

Beginning with Cilium 1.6, Kubernetes CRD-backed security identities can be
used for smaller clusters. Along with other changes in 1.6, this allows
kvstore-free operation if desired. It is possible to migrate identities from an
existing kvstore deployment to CRD-backed identities. This minimizes
disruptions to traffic as the update rolls out through the cluster.

Migration
~~~~~~~~~

When identities change, existing connections can be disrupted while Cilium
initializes and synchronizes with the shared identity store. The disruption
occurs when new numeric identities are used for existing pods on some instances
and others are used on others. When converting to CRD-backed identities, it is
possible to pre-allocate CRD identities so that the numeric identities match
those in the kvstore. This allows new and old Cilium instances in the rollout
to agree.

There are two ways to achieve this: you can either run a one-off ``cilium preflight migrate-identity`` script
which will perform a point-in-time copy of all identities from the kvstore to CRDs (added in Cilium 1.6), or use the "Double Write" identity

> allocation mode which will have Cilium manage identities in both the kvstore and CRD at the same time for a seamless migration (added in Cilium 1.16).

Reviewer 2: `s/1.16/1.17/`

      $ kubectl exec -n kube-system cilium-1234 -- cilium-dbg endpoint list # existing etcd-backed endpoints

Clearing CRD identities
$$$$$$$$$$$$$$$$$$$$$$$

If a migration has gone wrong, it possible to start with a clean slate. Ensure that no Cilium instances are running with ``--identity-allocation-mode=crd`` and execute:

.. code-block:: shell-session

      $ kubectl delete ciliumid --all

.. _double_write_migration:

Migration with the "Double Write" identity allocation mode

> ##########################################################

Reviewer 2: I'd suggest adding a note that this is a beta feature, and users may need extra care. This can be done with:

```
.. include:: ../beta.rst
```

> 
> 
> Example migration
> $$$$$$$$$$$$$$$$$

Reviewer 3: In Cilium docs we use `~` for this level of header. Please follow the same (same feedback below):

```suggestion
~~~~~~~~~~~~~~~~~
```
